### DISCUSSION TITLE
Midterm solution discussion

### DISCUSSION THREAD

**Discussion by MichaelPlug** [erodola/DLAI-s2-2022#24] [General]
Hello, I'm opening this discussion to talk about midterm exercises. Half of the students who participated in the midterm obtained an insufficient evaluation although the 5 extra-points.

I notice that Professor is not used to publish solutions of midterms or exams, so would be very useful if we collaborate to write it ourself. It would be a good thing if we used this space to publish our solution and discuss about it.  If this experiment would be reach the goal and be useful, in the future we can do the same thing with past exams.

I'm sharing with you a google docs like the midterm exam, so you'll can edit it adding your solution, or comment on the solution of others.

https://docs.google.com/document/d/1pCb1jEelUlM32UoS9gZld-VVEkMUnVK4/edit?usp=sharing&ouid=102447583511796840033&rtpof=true&sd=true

**Comment by erodola** [comment]
In the next few posts we'll copy-paste what we believe are very good answers to the midterm self-evaluation test. _These answers are taken directly from your own submissions_, but are kept anonymous.

Let's start with **Q1**: The applied regularization is assigning a loss that scales linearly w.r.t. the absolute value of the weights.  Since this regularization is applied on the boundaries of the range [-0.5,0.5] and because of the fact that a weight with a value close to 0 is giving no/a very little contribute to the network behaviour, most of the weights will be pushed around the boundary's values. The adopted loss can be expressed as max{10*|w|-5,0} where w=value of the weight.

**Comment by erodola** [comment]
**Q2**: Differently from SGD, here we are accumulating the quadratic value of the gradients inside grad_squared. We can recognize in this implementation something similar to the adaptive learning rate schema. At the first iteration , the gradient (and its square value) is computed. We recall that w.r.t. each of the parameter the value of the corresponding element of the gradient will be great if a small change of the parameter greatly affects the loss, on the other hand a small value of the gradient imply that a big change of the parameter slightly affects the loss value. This is exactly the purpose of this schema: scaling our learning rate by the square root of the square of the gradients, we will adopt a smaller learning rate along the steepest direction of the loss (in order to avoid overshooting) and a greater learning rate along the flattest direction of the loss (in order to boost the convergence). The main problem in this schema is (except the fact that during the first iteration the same learning rate will be used to update all the parameters in the model) that each element of np.sqrt(grad_squared) is a non decreasing quantity: because of that, the learning rate will unavoidably shrink to 0 as the learning goes on.

**Comment by lucmos** [comment]
**Q3**: 40 -> 20 -> batchnorm -> 10 -> dropout -> 3

hidden 1    : weights [40 * 20] + bias [20]    = 820
batchnorm : lambda [20] + beta [20]           = 40
hidden 2    : weights [20 * 10] + bias[10]     = 210
dropout     : weights [0]                                = 0
output        : weights [10 * 3] + bias [3]        = 33

total = 820 + 40 + 210 + 33 = 1103

**Reply by Theo-Clavel** [reply]
In the slide 17 of the batchnorm course, it says that the bias (of the linear layer) can be removed since it's ruled out by the mean substraction. So the right answer isn't 1083 ? (It's what I answered in my mid term exam)

**Reply by crisostomi** [reply]
The question explicitly stated that the MLP had the bias term, but in any case, the answer may be considered correct if you include your explanation.

**Reply by vitoc98** [reply]
Why is the input layer not considered? Shouldn't it have a certain number of weights dependent on the size of the input?

**Reply by ShakirKhurshid** [reply]
> Why is the input layer not considered? Shouldn't it have a certain number of weights dependent on the size of the input?

I think we have considered the input layer, it is the one with 40 neurons.

**Reply by vitoc98** [reply]
> > Why is the input layer not considered? Shouldn't it have a certain number of weights dependent on the size of the input?
> 
> I think we have considered the input layer, it is the one with 40 neurons.

And each one of the 40 neurons should have a parameter per every input value, meaning 40 * number_inputs additional parameters, right?

**Reply by lucmos** [reply]
Indeed the input is the one with 40 neurons, the first layer is represented by the arrow between 40 and 20.


You can try&play with the dimensions yourself directly in pytorch:
```python
from torch import nn

model = nn.Sequential(nn.Linear(40,20), 
                      nn.BatchNorm1d(20),  
                      nn.Linear(20,10), 
                      nn.Dropout(0.5), 
                      nn.Linear(10, 3))

print(sum(p.numel() for p in model.parameters() if p.requires_grad))
```

```
> 1103
```

**Reply by vitoc98** [reply]
Thank you for the info guys

**Comment by erodola** [comment]
**Q4:**

![image](https://user-images.githubusercontent.com/9395545/166723338-ce18b4db-f0c4-4e10-bee6-0f96b4d2ffae.png)

**Comment by erodola** [comment]
**Q5:** 

1. A Deep ReLU without batchnorm would struggle to aproximate the negative part of the identity, as this always gets fixed to zero
2. This would struggle even more as the data will be normalized after every step, skewing it away from an identity mapping.
3. Approximating identites is how VAEs are trained, so one should expect them to do a decent job at it
4. Should be possible

**Reply by giorgioskij** [reply]
When we say "Deep ReLU", do we mean that even the last layer has a ReLU activation? Because I think that this changes the reasoning for the answer.

**Reply by crisostomi** [reply]
This kind of reasoning is something you can (and should) write in the answer, maybe explaining what the answer would be in the two cases, i.e. with ReLU as the last layer and without.

**Comment by erodola** [comment]
**Q6:** The added pieces to the auto encoder (i.e. the ro and pi components) can actually be considered as 2 adversarial network (actually acting toward the same purpose): under the second term in the regularization l_lambda, the former network is trying to trasform the output of the encoder in such a way to resamble the output of the SPEC(X) (e.g. the frequenies of X). On the other hand, under the firtst term in l_lambda the latter (pi) network is trying to transform the output of SPEC(X) in such a way to resamble the the output of the encoder.Among the possible cases of usage of the above architecture that comes to my mind: 1) The model could be used as a Fouries Transformer : feeding a song X to the encoder we are able to extract its frequency description from the ro block, or feeding a set of frequencies to the pi block we are able to extract the approximated shape of the sound wave as the output of the decoder. 2) A powerful generative model that is both able to generate samples in the frequency space (sample from v -> feed to ro) or in the time space ((sample from v -> feed to D)

**Comment by crisostomi** [comment]
**Q7**: I'll try to implement a region proposal mechanism, I won't give the same weights to all the regions of the image but I would inspect only the most promising ones (0 padding the image outside the region of interest) so that the backgroud (if you don't want to preprocess the images clearing the noisy features) will have less importance, boxing at inference time only the most promising part of the images (using a certain accuracy metric). Also dropout, if not already used, could give a better generalization, allowing to not necessary taking into account every patch of the image.

**Comment by crisostomi** [comment]
**Q8**: This seems to me like a problem close to what is commonly defined as catastrophic forgetting. When a neural model is asked to learn different tasks in a sequential way, it always tries to optimize its weights at their absolute best for the current task (in this case, the current dataset) without any regard of which weights were important to perform the previous one (i.e: last month's dataset).

One naive way to counteract this problem would be to start with a pretty small model, and after each training session freeze the current weights, expand the model (adding initialized neurons to each layer) and only train those. Although very impractical, this would probably circumvent the problem in some way, but it's not scalable inifinitely and probably doesn't even lead to great performance.

A better way could be to introduce some sort of "memory", meaning that evey once in a while, during training on the new dataset, the model is fed some examples from the previous datasets. This could prevent the model from completely rewriting the weights that were useful for the previous data.

Lastly, a much better, more refined and scalable way could be to implement what was presented by Kirkpatrick et al. with the name of Elastic Weight Consolidation (EWC). Inspired by the human neurons, which become stiffer and more reluctant to change after having learned a new skill, or stored new knowledge, the paper proposes to introduce a formula similar to the mechanical equation of a spring for each neuron. This sort of elasitcity coefficient determines how easily can the weight of the neuron be changed, and is assigned based on how crucial that specific neuron was in the learning of the last skill. This would be for sure a quite complex solution to the problem, but very interesting at least.
